node/src/reactor/main_reactor.rs

```diff
@@ -960,6 +960,74 @@ impl reactor::Reactor for MainReactor {
                 self.storage.handle_event(effect_builder, rng, req.into()),
             ),
 
+            // This event gets emitted when we manage to read the era validators from the global
+            // states of an immediate switch block and its parent. Once that happens, we can check
+            // for the signs of any changes happening during the upgrade and register the correct
+            // set of validators in the validators matrix.
+            MainEvent::GotImmediateSwitchBlockEraValidators(
+                era_id,
+                parent_era_validators,
+                block_era_validators,
+            ) => {
+                // `era_id`, being the era of the immediate switch block, will be absent in the
+                // validators stored in the immediate switch block - therefore we will use its
+                // successor for the comparison.
+                let era_to_check = era_id.successor();
+                // We read the validators for era_id+1 from the parent of the immediate switch
+                // block.
+                let validators_in_parent = match parent_era_validators.get(&era_to_check) {
+                    Some(validators) => validators,
+                    None => {
+                        return fatal!(
+                            effect_builder,
+                            "couldn't find validators for era {} in parent_era_validators",
+                            era_to_check
+                        )
+                        .ignore();
+                    }
+                };
+                // We also read the validators from the immediate switch block itself.
+                let validators_in_block = match block_era_validators.get(&era_to_check) {
+                    Some(validators) => validators,
+                    None => {
+                        return fatal!(
+                            effect_builder,
+                            "couldn't find validators for era {} in block_era_validators",
+                            era_to_check
+                        )
+                        .ignore();
+                    }
+                };
+                // Decide which validators to use for `era_id` in the validators matrix.
+                let validators_to_register = if validators_in_parent == validators_in_block {
```

Comment:
That will compare both public key and weight. I wonder, if only the weight changes, do we consider it a validators update?

Reply:
I guess that for the purpose of determining weak vs strict finality correctly we would have to consider the weight as well.